src/blockchain/rpc.rs

```diff
@@ -495,6 +330,486 @@ fn list_wallet_dir(client: &Client) -> Result<Vec<String>, Error> {
+
+        Ok(self)
+    }
+
+    /// We want to filter out conflicting transactions.
+    /// Only accept transactions that are already confirmed, or existing in mempool.
+    fn _filter_tx(client: &Client, res: GetTransactionResult) -> Option<GetTransactionResult> {
+        if res.info.confirmations > 0 || client.get_mempool_entry(&res.info.txid).is_ok() {
+            Some(res)
+        } else {
+            debug!("tx filtered out: {}", res.info.txid);
+            None
+        }
+    }
+
+    /// Calculates received amount from raw tx.
+    fn _received_from_raw_tx<D: BatchDatabase>(db: &D, raw_tx: &Transaction) -> Result<u64, Error> {
+        raw_tx.output.iter().try_fold(0_u64, |recv, txo| {
+            let v = if db.is_mine(&txo.script_pubkey)? {
+                txo.value
+            } else {
+                0
+            };
+            Ok(recv + v)
+        })
+    }
+
+    /// Calculates sent from raw tx.
+    fn _sent_from_raw_tx<D: BatchDatabase>(
+        &self,
+        db: &D,
+        raw_tx: &Transaction,
+    ) -> Result<u64, Error> {
+        raw_tx.input.iter().try_fold(0_u64, |sent, txin| {
+            let v = match self._previous_output(&txin.previous_output) {
+                Some(prev_txo) => {
+                    if db.is_mine(&prev_txo.script_pubkey)? {
+                        prev_txo.value
+                    } else {
+                        0
+                    }
+                }
+                None => 0_u64,
+            };
+            Ok(sent + v)
+        })
+    }
+
+    fn _previous_output(&self, outpoint: &OutPoint) -> Option<&TxOut> {
+        let prev_tx = self.txs.get(&outpoint.txid)?.transaction.as_ref()?;
+        prev_tx.output.get(outpoint.vout as usize)
+    }
+
+    // updates the db state's last_index for the given keychain (if larger than current last_index)
+    fn _update_last_index(&mut self, keychain: KeychainKind, index: u32) {
+        let mut updated = false;
+
+        self.last_indexes
+            .entry(keychain)
+            .and_modify(|last| {
+                if *last < index {
+                    updated = true;
+                    *last = index;
+                }
+            })
+            .or_insert_with(|| {
+                updated = true;
+                index
+            });
+
+        if updated {
+            debug!(
+                "db state: last_index updated for {}: {}",
+                keychain.as_byte(),
+                index
+            );
+            self.updated_last_indexes.insert(keychain);
+        }
+    }
+
+    /// Prepare db batch operations.
+    fn update_batch<D: BatchDatabase>(&self, batch: &mut D::Batch) -> Result<(), Error> {
+        let mut del_txs = 0_u32;
+
+        // delete stale txs from db
+        // stale = not retained
+        self.txs
+            .keys()
+            .filter(|&txid| !self.retained_txs.contains(txid))
+            .try_for_each(|txid| batch.del_tx(txid, false).map(|_| del_txs += 1))?;
```

Comment:
I don't think this will work, as usually `batch.del_tx` will return `Some` only if you're deleting something that has just been inserted in the batch.

Reply:
Fixed.

Reply:
Is it? You're still doing:

```
if batch.del_tx(txid, false)?.is_some() {
    debug!("deleting tx: {}", txid);
    del_txs += 1;
}
```

But my point is that `del_tx` will delete a transaction even when returning None...

Ping @afilini for an explanation, I don't remember why is the BatchDatabase doing so

Reply:
It's because the batch object is basically disconnected from the actual database, it's just a collection of operations that are then applied at once at the end

Reply:
@afilini @danielabrozzoni Thanks for pointing this out! I'll actually fix it now